### PR TITLE
add connection string support

### PR DESCRIPTION
**Pull request by arily**

### CODE REVIEW

File: packages/plugin-mongo/src/database.ts [hunk 1]
```diff
@@ -43,4 +49,17 @@ export default class MongoDatabase {
...
   }
+
+  connectionStringFromConfig() {
+    let mongourl = `${this.config.protocol}://`
+    if (this.config.username) mongourl += `${this.config.username}${this.config.password ? `:${this.config.password}` : ''}@`
+    mongourl += `${this.config.host}${this.config.port ? `:${this.config.port}` : ''}/${this.config.authDatabase ? this.config.authDatabase : this.config.name}`
+    if (this.config.connectionOption) {
+      // https://nodejs.org/api/url.html#url_new_urlsearchparams_obj this should be find but I got an complaint from TS
+      const params = new URLSearchParams(this.config.connectionOption)
+      params.sort()
```
Comment (undefined-moe) [inline packages/plugin-mongo/src/database.ts:60]: why it must be sorted?

File: packages/plugin-mongo/src/database.ts [hunk 2]
```diff
@@ -1,14 +1,20 @@
 import { MongoClient, Db, Collection } from 'mongodb'
 import { App, TableType } from 'koishi-core'
+import { URLSearchParams } from 'url'
```
Comment (undefined-moe) [inline packages/plugin-mongo/src/database.ts:3]: `url` should import before `mongodb`

**Review by undefined-moe — CHANGES_REQUESTED** [review]

File: packages/plugin-mongo/src/database.ts [hunk 3]
```diff
@@ -24,9 +30,9 @@ export default class MongoDatabase {
   }
 
   async start() {
-    let mongourl = `${this.config.protocol}://`
-    if (this.config.username) mongourl += `${this.config.username}:${this.config.password}@`
-    mongourl += `${this.config.host}:${this.config.port}/${this.config.name}`
+    let mongourl
+    if (this.config.uri) mongourl = this.config.uri
+    else mongourl = this.connectionStringFromConfig()
```
Comment (Shigma) [inline packages/plugin-mongo/src/database.ts:35]:
```ts
mongourl = this.config.uri || this.connectionStringFromConfig()
```

File: packages/plugin-mongo/src/database.ts [hunk 4]
```diff
@@ -1,14 +1,20 @@
...
   username?: string
   password?: string
   protocol?: string
   host?: string
   port?: number
-  name?: string
+  name?: string // database name
   prefix?: string
+  authDatabase?: string // default auth database
+  connectionOption?: {
```
Comment (Shigma) [inline packages/plugin-mongo/src/database.ts:14]: prefer `connectionOptions`

File: packages/plugin-mongo/src/database.ts [hunk 5]
```diff
@@ -1,14 +1,20 @@
...
+    [key: string]: string | number | [string | number]
+  }
+  uri?: string // connection string (will overwrite all configs except 'name' and 'prefix')
```
Comment (Shigma) [inline packages/plugin-mongo/src/database.ts:17]: prefer jsdoc:
```diff
- uri?: string // connection string (will overwrite all configs except 'name' and 'prefix')
+ /** connection string (will overwrite all configs except 'name' and 'prefix') */
+ uri?: string
```

File: packages/plugin-mongo/src/database.ts (hunk 4 above)
Comment (arily) [reply packages/plugin-mongo/src/database.ts]: 40e27e2d5df047b4a98bc3a2fd9e1f669d0921b9 renamed

File: packages/plugin-mongo/src/database.ts (hunk 3 above)
Comment (arily) [reply packages/plugin-mongo/src/database.ts]: 05b28d32ff10ea1fc2637fd87b06b68736997988 fixed

File: packages/plugin-mongo/src/database.ts (hunk 1 above)
Comment (arily) [reply packages/plugin-mongo/src/database.ts:58]: oh my bad it doesn't

File: packages/plugin-mongo/src/database.ts (hunk 5 above)
Comment (arily) [reply packages/plugin-mongo/src/database.ts]: eb1ac0353f0f8fc2febdeff8e124f504dfa97fdc fixed

File: packages/plugin-mongo/src/database.ts (hunk 2 above)
Comment (arily) [reply packages/plugin-mongo/src/database.ts]: ef817b055586b5cf9d02ea8d32902f21c119229e fixed

File: packages/plugin-mongo/src/database.ts (hunk 1 above)
Comment (arily) [reply packages/plugin-mongo/src/database.ts]: 02ff7d9caa6430baf9c6f979dd30042e9c922179 fixed

File: packages/plugin-mongo/src/database.ts (hunk 2 above)
Comment (arily) [reply packages/plugin-mongo/src/database.ts]: ef817b055586b5cf9d02ea8d32902f21c119229e fixed

File: packages/plugin-mongo/src/database.ts [hunk 6]
```diff
@@ -43,4 +48,16 @@ export default class MongoDatabase {
   stop() {
     return this.client.close()
   }
+
+  connectionStringFromConfig() {
+    let mongourl = `${this.config.protocol}://`
+    if (this.config.username) mongourl += `${this.config.username}${this.config.password ? `:${this.config.password}` : ''}@`
+    mongourl += `${this.config.host}${this.config.port ? `:${this.config.port}` : ''}/${this.config.authDatabase ? this.config.authDatabase : this.config.name}`
+    if (this.config.connectionOptions) {
```
Comment (undefined-moe) [inline packages/plugin-mongo/src/database.ts:56]: what about `config.params`

**Review by undefined-moe — CHANGES_REQUESTED** [review]
also see ci issues

**Comment by arily** [comment]
> also see ci issues

03ee30134b8d7326752199cf2ae07fb1dcf099f5 will this fix it ? (my first time writing ts)

File: packages/plugin-mongo/src/database.ts (hunk 6 above)
Comment (arily) [reply packages/plugin-mongo/src/database.ts:56]: https://docs.mongodb.com/manual/reference/connection-string/
connection string refers to the "params" as "options", and options.options/ config.options feels confusing to me, thus why I chosed connectionOptions.
I also prefer a shorter key, so I am not sure which to use.

**Comment by arily** [comment]
finally works.

File: packages/plugin-mongo/src/database.ts [hunk 7]
```diff
@@ -7,8 +8,12 @@ export interface Config {
   protocol?: string
   host?: string
   port?: number
-  name?: string
+  name?: string // database name
   prefix?: string
+  authDatabase?: string // default auth database
+  connectionOptions?: URLSearchParams | string | { [key: string]: string | (string)[] } | Iterable<[string, string]> | Array<[string, string]>
```
Comment (undefined-moe) [inline packages/plugin-mongo/src/database.ts:14]: @types/node is using `URLSearchParams | string | NodeJS.Dict<string | ReadonlyArray<string>> | Iterable<[string, string]> | ReadonlyArray<[string, string]>`

**Review by undefined-moe — CHANGES_REQUESTED** [review]

**Comment by arily** [comment]
fixed https://github.com/koishijs/koishi/pull/119/commits/f05eecdbbce01fb95e938ae115a984117a01f584

**Review by undefined-moe — APPROVED** [review]
LGTM

**Comment by undefined-moe** [comment]
PTAL @Shigma

**Comment by Shigma** [comment]
@arily I adjusted interface and typings to fix warning from typescript.

**Pull request merged**